### ISSUE TITLE
As a reader, I want to download an article as a formatted PDF.

### ISSUE THREAD

**Issue by thatbudakguy** [Princeton-CDH/startwords#22]
## testing notes
To generate a PDF, install weasyprint (`pip install weasyprint` is one option) and then use weasyprint with the url of the article and give it a filename, e.g.
```
weasyprint http://startwords/issues/1/my-article/ my-filename.pdf
```
In addition to testing how the content looks in the PDF, please also check:
- table of contents should be generated based on headings in the document
- pdf metadata (in Adobe Acrobat this can be viewed by going to File -> Properties)
- links in the document (esp. header and footer links and endnote references and return links)

## dev notes
this issue includes however we want to represent 3D content and zoomable IIIF images in the pdf.

See comments on #70 for specifics on margin and font sizes that are not correct.

**Comment by rlskoeser** [comment]
@gissoo please review updated versions of the PDF articles and provide feedback. 

(I find it difficult to debug the styles since I can't get a browser to render the print css the same way weasyprint does, and I don't have the style inspecting tools for PDF that I'm used to in the browser.)

[dbv.pdf](https://github.com/Princeton-CDH/startwords/files/4799475/dbv.pdf)
[body-as-data.pdf](https://github.com/Princeton-CDH/startwords/files/4799579/body-as-data.pdf)

**Comment by gissoo** [comment]
@rlskoeser Thank you for working on this!! 
This looks awesome!! I was able to test the pdfs in Adobe Illustrator and compare with the numbers and designs on Figma. 

A couple of points:
1. From the top and from the bottom of the page, there is a 0.645in (you can round) container, which contains the header/footer. This looks right, however right after this container ends, there is supposed to be a 0.1in spacing between the text and the header/footer. Right now, I don't see this space, the spacing is less between the header and the text, and more between the footer and the text. 
[Here are the guides](https://www.figma.com/file/MEt3gfKX4N3WTLPbLGBbGC/Startwords?node-id=659%3A57) if you would like to take a look again

2. The margin (the red container in figma) in the pdf you have is 1.281in, which is even wider than what I have in the guide 1.11in. This is good, please don't make it 1.11in. However, there is a problem I can't figure out why it's happening. The body text is still looking wider than the design. I see that in my design the body text is around 220px (3.056in) wide and the body text in your pdf is 426.75px (5.927 in). Can we make the text narrower to match the design? Don't understand how the margins are correct but the body text's width is not correct :D :exploding_head:

3. Looking at it now, the THEME text is really competing with the article title here (also on the article page) - I'll ask for opinions in the channel, it might be better if we make it smaller :(. 

4. Could you please reduce the spacing between the theme and the issue number? Maybe by 1-2px

5. For RM's article PDF the pull quote's leading/line height is not correct. It's supposed to be a line height of 32. I don't have a way of testing it on Illustrator. 

0.1041666667 in

**Comment by rlskoeser** [comment]
@gissoo I don't think I can reliably put 0.1in spacing between the text and the header/footer. I could add a margin to the top of the document and to headers and paragraphs, but I don't think there's any way to guarantee that the text flows in such a way that we get that spacing.

Can we include that spacing in the 0.645 top and bottom margins? 

I'll look into the other things you've identified.

**Comment by rlskoeser** [comment]
@gissoo how are you converting between pixels and inches? I tried setting body content width to 3.056in and that is way too narrow, it's much narrower than your designs. It looks to me that the body content is 397px wide in Figma, but when I try setting that width exactly it still doesn't look right to me. I used the sizes in your designs to generate a percentage, and based on that I came up with a body text width of 5.51in.

**Comment by rlskoeser** [comment]
Revised PDFs for review! Have adjusted header/footer sizes, body content width, spacing between theme and issue number, and line-height and margins for the pull quote.

(Still don't understand how the math doesn't seem to be working the way I expect or what width you actually want for the content!) Hopefully we're close now, though.

[dbv.pdf](https://github.com/Princeton-CDH/startwords/files/4821829/dbv.pdf)
[body-data.pdf](https://github.com/Princeton-CDH/startwords/files/4821830/body-data.pdf)

**Comment by gissoo** [comment]
> @gissoo I don't think I can reliably put 0.1in spacing between the text and the header/footer. I could add a margin to the top of the document and to headers and paragraphs, but I don't think there's any way to guarantee that the text flows in such a way that we get that spacing.
> 

- I suspected that too..

> Can we include that spacing in the 0.645 top and bottom margins?
> 

- Sure! If you try it and it messes things up then please don't worry about it. 

> I'll look into the other things you've identified.

**Comment by gissoo** [comment]
> @gissoo how are you converting between pixels and inches? I tried setting body content width to 3.056in and that is way too narrow, it's much narrower than your designs. It looks to me that the body content is 397px wide in Figma, but when I try setting that width exactly it still doesn't look right to me. I used the sizes in your designs to generate a percentage, and based on that I came up with a body text width of 5.51in.
>

- @rlskoeser I'm sorry looking at it again now on Figma and Illustrator I don't know where I got the 220px from either. So 5.51in makes sense! Looking at your revised pdfs now I still the body width as 426.75px (5.927 in). I don't know why it's not changing to 5.51in ....What if you make it smaller than 5.51? Does it change? How about trying 5in? 

You've defined the margins, right? What if you don't define them? I think what might be happening is that it's restricting the size of the body width based on the defined margins (left and right)....Because the margins match the design as I mentioned...it's the width that's not matching ...

**Comment by gissoo** [comment]
ah please ignore my last bullet ...The margins do not match the design which I said is fine ...The problem must be somewhere else

**Comment by gissoo** [comment]
@rlskoeser Thank you!! Everything else looks perfect!!!

**Comment by rlskoeser** [comment]
@gissoo I don't know what happened with the margins on the last version I asked you to look at. Here's a freshly created PDF with the latest code (renamed the file to distinguish) — I just checked with the rulers in Adobe Acrobat and the body content width looks pretty close to  5.5 inches.

[dbv-latest.pdf](https://github.com/Princeton-CDH/startwords/files/4832022/dbv-latest.pdf)

**Comment by gissoo** [comment]
@rlskoeser thank you so much, you're awesome!!!! 

this looks amazing!!!
Just two points:
- On the last page of the pdf I don't know what is happening, all the visual assets have gathered :D 
![Screen Shot 2020-06-25 at 2.22.00 PM.png](https://images.zenhubusercontent.com/5bf307a69d595b459ec91bb1/6a6e0ef4-f084-4f35-be06-c24780fcece6)

- I just noticed the section headers' font sizes (as well as the "Goals" .."Description") are 15 with a line height of 18 I think, in the designs their font size is 14 with a line height of 21. (Also, not sure if you are still okay with/planning to apply [how the section headers are styled in the designs here](https://www.figma.com/file/MEt3gfKX4N3WTLPbLGBbGC/Startwords?node-id=462%3A149) (having narrower width) – if it's not possible, it's okay). I think applying

**Comment by rlskoeser** [comment]
@gissoo yes, Grant pointed out the endnote styles problem too — I'm working on a fix but now I can't get the numbers to display in the right place! 😬  

I missed the section header font sizes, sorry. Thanks for catching. The DBV-specific styles aren't in place yet and I wasn't sure how much to do on things related to that. Would the narrow width apply even on articles with no icons for the sections, or is this a DBV-specific style?

**Comment by gissoo** [comment]
@rlskoeser Sorry that you're going through this. :( 

- I suspected that too ...It's really hard to tell for future articles ...The DBV essay has too many headers, so this is helping it with readability. I do think this style in general will help with long headers (I just don't know which header level it will be applied for other times/future cases). I'm leaning towards saying yes, it will apply for headers in other articles, and potentially any level of header within the body (because its goal is to make long headers readable). (except the article title which has its own fixed width).

**Comment by rlskoeser** [comment]
Ok @gissoo — revised pdf with cleaned up end notes styles and adjusted header styles. I set the line height that I saw in Figma (which is what you said), but it doesn't look quite right in the PDF. Let me know if you want me to adjust.

[dbv-latest.pdf](https://github.com/Princeton-CDH/startwords/files/4833415/dbv-latest.pdf)

**Comment by gissoo** [comment]
@rlskoeser Thank you!! The width is perfect. For some reason the size of both headers is still showing as 15. Could you increase the line height by 2 more points? It's still showing the same number to me. (18 on Adobe Illustrator, and 0 on PDF)

**Comment by rlskoeser** [comment]
I think the weasyprint header styles were conflicting with the local styles. I made the local styles more specific and I think things look better now.

[dbv-latest.pdf](https://github.com/Princeton-CDH/startwords/files/4833522/dbv-latest.pdf)

**Comment by rlskoeser** [comment]
I noticed the end note font sizes were still not correct (another style conflict / priority issue, I think.) Here's an updated PDF that corrects it.
[dbv-latest.pdf](https://github.com/Princeton-CDH/startwords/files/4846517/dbv-latest.pdf)

**Comment by gissoo** [comment]
@rlskoeser This looks great!! The only things that would help to revise is, if possible:
- can you change the THEME font size to 11? It's now 12 in my PDF view?
- can you change the authors' text font weight from light to regular? 
- Because the body text is now set at 9, would you please change the authors and the date and the doi to be 11 as well? 

Thank you!!

**Comment by rlskoeser** [comment]
@gwijthoff Gissoo and I are still fine-tuning a few of the PDF styles, but please proceed with testing and review whenever you have time. Use weasyprint to create a PDF of both article from the startwords url and check them over thoroughly, including the items I've listed in the testing notes on this issue (table of contents, internal links, metadata).

**Comment by rlskoeser** [comment]
@gissoo generated a new PDF from the production site with the style revisions you requested. (Fine to review later in the week when you're looking at other things)
[dbv-prod-2020-07-06.pdf](https://github.com/Princeton-CDH/startwords/files/4880544/dbv-prod-2020-07-06.pdf)

**Comment by gissoo** [comment]
@rlskoeser Thank you, Rebecca!! When I test the doc in the PDF edit view I see that the THEME text is now 8.25 :( don't know how this is happening, it's much smaller now than it was (before this it was only 1 size up than it is supposed to be, I think). The same is happening to the doi text. Could you increase them by a bit? 

The author's font weight is great now thank you!!  

This might be easiest: Would you please also make the issue number, theme, authors, issue date, and the doi the same size? whatever they are it would be great if they are the same size (just looking at the designs now) – it would be great if they are all between 10 or 11. Sorry that this keeps happening.

**Comment by gwijthoff** [comment]
@rlskoeser Gissoo's comments on text size aside, everything in the PDF looks great to me! I tested weasyprint using the production site and running locally on the `feature/pdf` branch (lost track of which I was supposed to test!). FWIW, when running locally, header font sizes (especially issue theme) are larger than the production site.

**Comment by gwijthoff** [comment]
Here's the metadata for the pdf generated by the `feature/pdf` branch:

<img width="263" alt="Screen Shot 2020-07-07 at 5 41 47 PM" src="https://user-images.githubusercontent.com/5580840/86846705-40dca800-c079-11ea-9448-8c857a8fce14.png">

And here's the metadata generated from the current production site:

<img width="262" alt="Screen Shot 2020-07-07 at 5 43 08 PM" src="https://user-images.githubusercontent.com/5580840/86846776-54880e80-c079-11ea-9035-68079ee2c05e.png">

Love the inclusion of "A research periodical irregularly published by the Center for Digital Humanities at Princeton." Would it be too wordy to include that after "Startwords Issue 1, January 2020."?

**Comment by rlskoeser** [comment]
@gissoo I finally figured out what's going on here! You gave me font sizes without units, and I assumed pixels because that's the font size unit I see in Figma — but I think the Adobe Acrobat text editor is showing font size in points! (It doesn't display a unit either 🙄 )

Please let me know how you want me to adjust the fonts based on hits information — either by providing specific values and units for what you want adjusted, or by letting me know how to calculate them correctly from what's in Figma.  (I wonder if using the pixel sizes in Figma and converting to `rem` would fix this?)

**Comment by gissoo** [comment]
@rlskoeser Thank you for all of your work on this. As we agreed, I'm writing here all the font sizes in pixels from now on and only specifying those that need to be revise. I'm looking at the latest pdf you provided in this thread, the one called "dbv-prod-2020-07-06.pdf". 
1. The THEME in the header, issue number, authors, issue date, and the doi need to be 14px. 
2. The authors font weight needs to be "regular"

I think these are the only changes that were pending.

**Comment by rlskoeser** [comment]
@gissoo have revised the font settings you listed. Here are PDFs of the two articles generated with the change:

[2020-07-20-bodydata.pdf](https://github.com/Princeton-CDH/startwords/files/5104092/2020-07-20-bodydata.pdf)
[2020-08-20-dbv.pdf](https://github.com/Princeton-CDH/startwords/files/5104093/2020-08-20-dbv.pdf)

**Comment by gissoo** [comment]
@rlskoeser thank you!!! This looks awesome!!! The only thing that I just want to make sure, is the body text 12 pixels and the leading 18? The spacing seems like tighter than in [the designs ](https://www.figma.com/file/MEt3gfKX4N3WTLPbLGBbGC/Startwords?node-id=462%3A48) – if they match, then don't worry about it, it's fine, and I can close the issue then.

**Comment by rlskoeser** [comment]
@gissoo I just inspected the print css styles in the browser (Chrome lets you emulate print rendering), and it's showing the body text as having a font size of 12 pixels and a line height of 18 pixels.

I also inspected the PDFs I shared with you in Adobe Acrobat, and it shows the font size as 9 and the spacing as 13.5. I just checked with a point to pixel converter and this seems correct.

**Comment by gissoo** [comment]
Thank you for all of your work on this, @rlskoeser!!
